Source/WebCore/Modules/model-element/HTMLModelElement.cpp

```diff
@@ -231,40 +232,67 @@ void HTMLModelElement::didAttachRenderers()
 
 void HTMLModelElement::dataReceived(CachedResource& resource, const SharedBuffer& buffer)
 {
-    ASSERT_UNUSED(resource, &resource == m_resource);
-    m_data.append(buffer);
+    if (&resource == m_resource)
+        m_data.append(buffer);
+#if ENABLE(MODEL_PROCESS)
+    else if (&resource == m_environmentMapResource)
+        m_environmentMapData.append(buffer);
+#endif
+    else
+        ASSERT_NOT_REACHED();
 }
 
 void HTMLModelElement::notifyFinished(CachedResource& resource, const NetworkLoadMetrics&, LoadWillContinueInAnotherProcess)
 {
-    auto invalidateResourceHandleAndUpdateRenderer = [&] {
-        m_resource->removeClient(*this);
-        m_resource = nullptr;
+    auto invalidateResourceHandleAndUpdateRenderer = [&](CachedResourceHandle<CachedRawResource>& resourceHandle) {
+        resourceHandle->removeClient(*this);
+        resourceHandle = nullptr;
 
         if (CheckedPtr renderer = this->renderer())
             renderer->updateFromElement();
```

Comment:
Do we really need to do this renderer update for the environment map case? I didn't think we need to.

I think it may be clearer to just have two helper methods to handle the load completion of model resource and the environment map resource respectively. This way this `invalidateResourceHandleAndUpdateRenderer` block wouldn't need to take in a resource handle either.

Then this method will call the correct helper method depending on the `resource` identity.

Reply:
Added two helper methods, one for model resource and one for environment map resources. Model's resource original code kept without changes. This method calls corresponding helpers.